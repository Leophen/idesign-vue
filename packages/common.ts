import _ from 'lodash'
import {
  Component,
  getCurrentInstance,
  Slots,
  VNode,
} from 'vue'

// 通用位置类型
export type placementType =
  | 'top'
  | 'top-left'
  | 'top-right'
  | 'bottom'
  | 'bottom-left'
  | 'bottom-right'
  | 'left'
  | 'left-top'
  | 'left-bottom'
  | 'right'
  | 'right-top'
  | 'right-bottom'

// 通用触发类型
export type triggerType = 'hover' | 'click' | 'context-menu'

// 转换 width/height/left/... 样式值
export const turnValue = (val: string | number | undefined) => {
  if (!val) {
    return
  }
  if (_.isNumber(val as number)) {
    return `${val}px`
  }
  return val
}

// const _stateValue = ref(defaultValue)
// const innerValue = computed(() => value ?? _stateValue.value)

/**
 * 渲染 default slot，获取子组件 VNode。处理多种子组件创建场景
 * @returns {function(childComponentName: string, slots: Slots): VNode[]}
 * @param childComponentName
 * @param slots
 * @example const getChildByName = useChildComponentSlots()
 * @example getChildComponentByName('xxItem')
 */
export const useChildComponentSlots = () => {
  const instance = getCurrentInstance()
  return (childComponentName: string, slots?: Slots): VNode[] => {
    if (!slots) {
      slots = instance?.slots
    }
    const content = slots?.default?.() || []

    return (
      content
        .map((item: VNode) => {
          if (item.children && Array.isArray(item.children))
            return item.children
          return item
        })
        .flat()
        // @ts-ignore
        .filter((item: VNode) =>
          (item.type as Component).name?.endsWith(childComponentName)
        ) as VNode[]
    )
  }
}
