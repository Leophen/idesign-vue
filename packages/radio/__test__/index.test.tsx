import { vi } from 'vitest';
import { Radio, RadioGroup } from '../index';
import { render } from '../../common'

describe('Radio 组件测试', () => {
  it('create', () => {
    const wrapper = render(<Radio />);
    expect(wrapper.element).toMatchSnapshot();
  });

  it('checked & children', () => {
    const wrapper = render(<Radio modelValue={true}>广州</Radio>);
    expect(wrapper.find('.i-radio').classes()).toContain('i-radio-is-checked');
    expect(wrapper.find('.i-radio__label').text()).toBe('广州');
  });

  it('defaultChecked', () => {
    const wrapper = render(<Radio defaultChecked={true}>广州</Radio>);
    expect(wrapper.find('.i-radio').classes()).toContain('i-radio-is-checked');
  });

  it('onChange', () => {
    const changeFn = vi.fn();
    const wrapper = render(<Radio onChange={changeFn} />)
    wrapper.find('.i-radio__former').trigger('change');
    expect(changeFn).toBeCalledTimes(1);
  });

  it('disabled', () => {
    const fn = vi.fn();
    const wrapper = render(<Radio disabled={true} onChange={fn} />)
    expect(wrapper.find('.i-radio').classes()).toContain('i-radio-is-disabled');
    wrapper.find('.i-radio__former').trigger('change');
    expect(fn).toBeCalledTimes(0);
  });

  it('type', () => {
    const wrapper = render(<Radio type="radio-button" />);
    expect(wrapper.classes()).toContain('i-radio-button');
  });

  it('size', () => {
    const wrapper = render(<Radio type="radio-button" size="small" />);
    expect(wrapper.classes()).toContain('i-radio-button--size-small');
  });
});

describe('RadioGroup 组件测试', () => {
  it('selected', () => {
    const wrapper = render(
      <RadioGroup modelValue='gz'>
        <Radio value="gz">广州</Radio>
      </RadioGroup>,
    );
    expect(wrapper.find('.i-radio').classes()).toContain('i-radio-is-checked');
  });

  it('defaultSelected', () => {
    const wrapper = render(
      <RadioGroup defaultSelected='sz'>
        <Radio value="gz">广州</Radio>
      </RadioGroup>,
    );
    expect(wrapper.find('.i-radio').classes()).toContain('i-radio-is-checked');
  });

  it('onChange', () => {
    const changeFn = vi.fn();
    const wrapper = render(
      <RadioGroup onChange={changeFn}>
        <Radio value="gz">广州</Radio>
      </RadioGroup>,
    );
    wrapper.find('.i-radio__former').trigger('change');
    expect(changeFn).toHaveBeenCalled();
  });

  it('disabled', () => {
    const changeFn = vi.fn();
    const wrapper = render(
      <RadioGroup disabled onChange={changeFn}>
        <Radio value="gz">广州</Radio>
      </RadioGroup>,
    );
    expect(wrapper.find('.i-radio').classes()).toContain('i-radio-is-disabled');
    wrapper.findComponent(Radio).trigger('click');
    expect(changeFn).toBeCalledTimes(0);
  });

  it('size', () => {
    const wrapper = render(
      <RadioGroup type="radio-button" size="small">
        <Radio value="gz">广州</Radio>
      </RadioGroup>,
    );
    expect(wrapper.find('.i-radio-button').classes()).toContain('i-radio-button--size-small');
  });
});

