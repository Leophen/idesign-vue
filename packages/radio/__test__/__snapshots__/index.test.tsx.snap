// Vitest Snapshot v1

exports[`Radio ç»„ä»¶æµ‹è¯• > create 1`] = `
<label
  class="i-radio i-radio--size-medium"
>
  <input
    class="i-radio__former"
    readonly=""
    type="radio"
  />
  <span
    class="i-radio__input"
  />
  <span
    class="i-radio__label"
  >
    <!---->
  </span>
</label>
`;
